program/src/processor.rs

```diff
@@ -475,10 +554,61 @@ fn process_harvest_rewards(program_id: &Pubkey, accounts: &[AccountInfo]) -> Pro
 /// Processes an SPL Transfer Hook Interface
 /// [ExecuteInstruction](https://docs.rs/spl-transfer-hook-interface/latest/spl_transfer_hook_interface/instruction/struct.ExecuteInstruction.html).
 pub fn process_spl_transfer_hook_execute(
-    _program_id: &Pubkey,
-    _accounts: &[AccountInfo],
-    _amount: u64,
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    transfer_amount: u64,
 ) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    let source_token_account_info = next_account_info(accounts_iter)?;
+    let mint_info = next_account_info(accounts_iter)?;
+    let destination_token_account_info = next_account_info(accounts_iter)?;
+    let _source_owner_info = next_account_info(accounts_iter)?;
+    let _extra_metas_info = next_account_info(accounts_iter)?;
+    let holder_rewards_pool_info = next_account_info(accounts_iter)?;
+    let source_holder_rewards_info = next_account_info(accounts_iter)?;
+    let destination_holder_rewards_info = next_account_info(accounts_iter)?;
+
+    let current_accumulated_rewards_per_token = {
+        check_pool(program_id, mint_info.key, holder_rewards_pool_info)?;
+        let pool_data = holder_rewards_pool_info.try_borrow_data()?;
+        let pool_state = bytemuck::try_from_bytes::<HolderRewardsPool>(&pool_data)
+            .map_err(|_| ProgramError::InvalidAccountData)?;
+        pool_state.accumulated_rewards_per_token
+    };
+
+    // Update the source holder rewards account.
+    //
+    // For the source - since it was just debited - the transfer amount
+    // will be added back to calculate the rewards share before the
+    // transfer.
+    update_holder_rewards_for_transfer_hook(
+        program_id,
+        mint_info.key,
+        source_token_account_info,
+        source_holder_rewards_info,
+        current_accumulated_rewards_per_token,
+        |amount| {
+            amount
+                .checked_add(transfer_amount)
+                .ok_or(ProgramError::ArithmeticOverflow)
+        },
+    )?;
+
+    // Update the destination holder rewards account.
+    //
+    // For the destination - since it was just credited - the transfer
+    // amount will be subtracted to calculate the rewards share before
+    // the transfer.
+    update_holder_rewards_for_transfer_hook(
+        program_id,
+        mint_info.key,
+        destination_token_account_info,
+        destination_holder_rewards_info,
+        current_accumulated_rewards_per_token,
+        |amount| Ok(amount.saturating_sub(transfer_amount)),
```

Comment:
nit: is there a reason to do saturating in this closure and checked with the other? We should be able to do checked with both of them, right?

Reply:
Yeahhh I kind of let this one slide because Token-2022 is going to check the amounts already, so it should never really even go over. For some reason I have a bias against "checked" it appears. 👻